In [ ]:
! pip install -q pyomo

In [ ]:
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
# create a model
model = ConcreteModel()

In [ ]:
N = 12 # number of decision variables
M = 5 # number of constraints

In [ ]:
obj_coef_c = np.array([0.56, 1.721, -9.6, -2.85, 41.5, 33.05, -0.75, 20.5, -0.35, -31.02, 50.65, 6.125])

In [ ]:
constr_coef_A = ([[1, -1, 0, 0, 2, -2, 1, 0, 0, -1, 1, -2],
                 [1, 0, 0, -1, 0, -2, 0, -1, 1, 0, 0, -1],
                 [-1, 0, 0, 1, -1, 2.5, -3.5, 0, 2, -1, 1, 0],
                 [0, 1, -0.45, 0, -2.78, 0, -3.4, 0, -1, 1, -1, 0.25],
                 [0, -1, 1, 0, 1, 0, 1, 3, -2, 0, 1, 0]])
constr1_rhs_b = ([88.5, 35.4, 47.9, 41.43, 22.7])

In [ ]:
lower_bound = np.array([0, 0, -np.inf, -np.inf, -34, 0, -67, 0, -23, -5, 6, 1])
upper_bound = np.array([12, 503, 4, np.inf, np.inf, 2, 20, 55, 23, 15, np.inf, np.inf])

In [ ]:
col_indices = np.arange(N)
row_indices = np.arange(M)
print(col_indices)


[ 0  1  2  3  4  5  6  7  8  9 10 11]


In [ ]:
model.x = Var(col_indices)

In [ ]:
model.constraints = ConstraintList()

In [ ]:
for i in row_indices:
  model.constraints.add(sum(constr_coef_A[i][j]*model.x[j] for j in col_indices) <= constr1_rhs_b[i])

In [ ]:
for j in col_indices:
  model.x[j].setlb(lower_bound[j])
  model.x[j].setub(upper_bound[j])

In [ ]:
model.objective = Objective(expr = sum(obj_coef_c[j]*model.x[j] for j in col_indices), sense=minimize)

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   12 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}

1 Var Declarations
    x : Size=12, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :   0.0 :  None :  12.0 : False :  True :  Reals
          1 :   0.0 :  None : 503.0 : False :  True :  Reals
          2 :  None :  None :   4.0 : False :  True :  Reals
          3 :  None :  None :  None : False :  True :  Reals
          4 : -34.0 :  None :  None : False :  True :  Reals
          5 :   0.0 :  None :   2.0 : False :  True :  Reals
          6 : -67.0 :  None :  20.0 : False :  True :  Reals
          7 :   0.0 :  None :  55.0 : False :  True :  Reals
          8 : -23.0 :  None :  23.0 : False 

In [ ]:
!apt-get install -y -qq glpk-utils

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -1951.831
  Upper bound: -1951.831
  Number of objectives: 1
  Number of constraints: 6
  Number of variables: 13
  Number of nonzeros: 38
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.012552022933959961
# ----------------------------------------------------------
#   Solution Information
# --------------------------

In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  -1951.8309999999997

Decision Variables
x[ 0 ] =  12.0
x[ 1 ] =  0.0
x[ 2 ] =  4.0
x[ 3 ] =  119.82
x[ 4 ] =  -34.0
x[ 5 ] =  0.0
x[ 6 ] =  20.0
x[ 7 ] =  0.0
x[ 8 ] =  -7.46000000000002
x[ 9 ] =  15.0
x[ 10 ] =  6.0
x[ 11 ] =  1.0

Constraints
constraints : Size=5
    Key : Lower : Body                : Upper
      1 :  None :               -47.0 :  88.5
      2 :  None : -116.28000000000002 :  35.4
      3 :  None :   47.89999999999995 :  47.9
      4 :  None :   41.43000000000002 : 41.43
      5 :  None :   10.92000000000004 :  22.7


Answer:

1. Optimal values of decision variables:=
$ \\ x_0 =  12.0 \\
x_1 =  0.0 \\
x_2 =  4.0 \\
x_3=  119.82 \\
x_4 =  -34.0 \\
x_5 =  0.0 \\
x_6 =  20.0 \\
x_7 =  0.0 \\
x_8 =  -7.46000000000002 \\
x_9 =  15.0 \\
x_{10} =  6.0 \\
x_{11} =  1.0
 $ 
2.Optimal cost:  -1951.8309999999997 $ \\ $
3.Activeness of Constraints: $ \\ $

(a) constraint 1st is not active since LHS =   -47.0 which is not equal to RHS. $ \\  $

(b)constraint 2nd is also not active since LHS = -116.28000000000002 which is less than RHS. 

(c) constraint 3rd is active constraint since LHS = 47.89999999999995 which is approximately equal to RHS.

(d) constraint 4th is active constraint since LHS =  41.43000000000002 which is equal to RHS.

(e) constraint 5th is not active constraint since LHS = 10.92000000000004 which is not equal to RHS.